src/composer.ts

```diff
@@ -303,18 +316,95 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
+    ): Composer<FilteredMatchContext<C, string, ":text" | ":caption">>;
+    hears(
+        trigger: MaybeNonEmptyArray<RegExp>,
+        ...middleware: Array<
+            Middleware<
+                FilteredMatchContext<C, RegExpMatchArray, ":text" | ":caption">
+            >
+        >
+    ): Composer<
+        FilteredMatchContext<C, RegExpMatchArray, ":text" | ":caption">
+    >;
+    hears(
+        trigger: MustMixedNonEmptyArrayType<string, RegExp>,
+        ...middleware: Array<
+            Middleware<
+                FilteredMatchContext<
+                    C,
+                    string | RegExpMatchArray,
+                    ":text" | ":caption"
+                >
+            >
+        >
+    ): Composer<
+        FilteredMatchContext<C, string | RegExpMatchArray, ":text" | ":caption">
+    >;
+    hears(
+        trigger: MaybeNonEmptyArray<string | RegExp>,
+        ...middleware: Array<
+            Middleware<
+                FilteredMatchContext<
+                    C,
+                    string | RegExpMatchArray,
+                    ":text" | ":caption"
+                >
+            >
+        >
+    ) {
+        const triggers = toNonEmptyArray(trigger);
+        const triggerFns = triggers.map((t) => triggerFn(t));
+        if (isNonEmptyStringArray(triggers)) {
+            return this.on([":text", ":caption"]).filter(
+                (
+                    ctx,
+                ): ctx is FilteredMatchContext<
+                    C,
+                    string,
+                    ":text" | ":caption"
+                > => {
+                    const msg = ctx.message ?? ctx.channelPost;
+                    const txt = msg.text ?? msg.caption;
+                    return match(ctx, txt, triggerFns);
+                },
+                ...middleware,
+            );
+        } else if (isNonEmptyRegExpArray(triggers)) {
+            return this.on([":text", ":caption"]).filter(
+                (
+                    ctx,
+                ): ctx is FilteredMatchContext<
+                    C,
+                    RegExpMatchArray,
+                    ":text" | ":caption"
+                > => {
+                    const msg = ctx.message ?? ctx.channelPost;
+                    const txt = msg.text ?? msg.caption;
+                    return match(ctx, txt, triggerFns);
+                },
+                ...middleware,
+            );
+        } else {
+            return this.on([":text", ":caption"]).filter(
+                (
+                    ctx,
+                ): ctx is FilteredMatchContext<
+                    C,
+                    string | RegExpMatchArray,
+                    ":text" | ":caption"
+                > => {
+                    const msg = ctx.message ?? ctx.channelPost;
+                    const txt = msg.text ?? msg.caption;
+                    return match(ctx, txt, triggerFns);
+                },
+                ...middleware,
+            );
+        }
```

Comment:
I don't think we should replicate the same code so many times. If TypeScript sucks at unifying things here, then we should type cast the arguments explicitly.

Reply:
Agreed. Trying to avoid casting right now though.